coldataext: remove a wrapper and use tree.Datum directly

yuzefovich opened the pull request:
**coldataext: propagate evalCtx directly for casts**

This commit removes the reliance on propagating `datumVec` objects to
supply the cast function with an eval context. All casts in the datum
land have been refactored to use `tree.PerformCast` for simplicity.

Note that the eval context is not removed from the `datumVec` because it
is used by `CompareDatum` method which is a lot more common (and, thus,
would require more plumbing to get rid off).

Release note: None

**coldataext: remove a wrapper and use tree.Datum directly**

Previously, we had separate `coldataext.Datum` wrapper around
`tree.Datum`. I don't remember what was the reasoning behind introducing
it (probably so that we could define methods with the wrapper as the
receiver in `coldataext` package), but it seems unnecessary. It also has
some performance cost because on every `DatumVec.Get` we are currently
allocating a new object on the heap. This commit removes the wrapper in
favor of working with `tree.Datum`s directly. This work was prompted by
looking at some profiles around the memory accounting for datum-backed
types in the cFetcher.

Release note: None

yuzefovich committed:
coldataext: remove a wrapper and use tree.Datum directly

Previously, we had separate `coldataext.Datum` wrapper around
`tree.Datum`. I don't remember what was the reasoning behind introducing
it (probably so that we could define methods with the wrapper as the
receiver in `coldataext` package), but it seems unnecessary. It also has
some performance cost because on every `DatumVec.Get` we are currently
allocating a new object on the heap. This commit removes the wrapper in
favor of working with `tree.Datum`s directly. This work was prompted by
looking at some profiles around the memory accounting for datum-backed
types in the cFetcher.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68249)
<!-- Reviewable:end -->

yuzefovich committed:
coldataext: propagate evalCtx directly for casts

This commit removes the reliance on propagating `datumVec` objects to
supply the cast function with an eval context. All casts in the datum
land have been refactored to use `tree.PerformCast` for simplicity.

Note that the eval context is not removed from the `datumVec` because it
is used by `CompareDatum` method which is a lot more common (and, thus,
would require more plumbing to get rid off).

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

This is a really nice refactor.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68249#-:-Mft47b7A4GBOW9ngMa0:b-ck5wg6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
